### DISCUSSION TITLE
Handling half-closed HTTP/2 streams gracefully.

### DISCUSSION THREAD

**Discussion by lovelydinosaur** [encode/httpcore#750] [General]
Pull request #641 added support for handling HTTP/1.1 half-closed connections gracefully.

In plainer language... when the client sends a large request, the server can close the request pre-emptively "nope this is too big not going to accept it" and still send a valid response. "nope that was too big, I didn't accept it". It's more user friendly if the client can return that response rather than just raising an exception at the point the request was closed.

We probably want to handle the same case for HTTP/2, but... it'd be really helpful to be able to demonstrate it in the wild. Can anyone point at an example 413 request against an HTTP/2 server that we could use to verify this as a requirement?

**Comment by karpetrosyan** [comment]
Code
``` python
import httpcore

pool = httpcore.ConnectionPool(http1=False, http2=True)
response = pool.request("GET", "http://127.0.0.1:80", content=b'1'*10_000_000)  # fails
response = pool.request("GET", "http://127.0.0.1:80", content=b'1'*10)  # ok
```

Nginx conf
```
http 
{
      server {
          listen 80 http2;
      }
  }
events {}
```

**Comment by luchagen** [comment]
Good morning,
We experienced this in the context of QA automated testing for an internal file upload solution.
It would seem that in this case, there is no exception to catch, as this was the scenario :

- client posts headers to the server
- server decides to reject based on the content-length
- client wants to start sending the request body, and waits for a WINDOW_UPDATE frame (HTTP2Connection._wait_for_outgoing_flow )
- server sends response headers, data  and stream reset while client is waiting for WINDOW_UPDATE, but the former are ignored.
- WINDOW_UPDATE is never received , so client waits until timeout or server side graceful termination of the connection


I had some success adding a check and subsequent raise on the events stored in HTTP2Connection._events[stream_id] in the loop of _wait_for_outgoing_flow as a workaround.

I will try to find a publicly accessible example when I can.
Best Regards,
Luc.
